test/metabase/db/util_test.clj

```diff
@@ -0,0 +1,134 @@
+  ;; Using Setting is perfect because it has only two required fields - (the primary) key & value (with no constraint).
+  ;;
+  ;; In the `:key` case using the `idempotent-insert!` rather than an `or` prevents the from application throwing an
+  ;; exception when there are race conditions. For `:value` it prevents us silently inserting duplicates.
+  ;;
+  ;; It's important to test both, as only the latter has a phantom read issue and thus requires serializable isolation.
+  (let [columns [:key :value]]
+    (doseq [search-col columns]
+      (testing (format "When the search column %s a uniqueness constraint in the db"
+                       (if (= :key search-col) "has" "does not have"))
+        (let [search-value   (str (random-uuid))
+              other-col      (first (remove #{search-col} columns))]
+          (try
+            ;; ensure there is no database detritus to trip us up
+            (t2/delete! Setting search-col search-value)
+
+            (let [threads 5
+                  latch   (CountDownLatch. threads)
+                  thunk   (fn []
+                            (mdb.u/select-or-insert! Setting {search-col search-value}
+                                                     (fn []
+                                ;; Make sure all the threads are in the mutating path
+                                (.countDown latch)
+                                (.await latch)
+                                {other-col (str (random-uuid))})))
+                  results (repeat-concurrently threads thunk)
+                  n       (count results)
+                  latest  (t2/select-one Setting search-col search-value)]
+
+              (case search-col
+                :key
+                (do (testing "every call returns the same row"
+                      (is (= #{latest} results)))
+
+                    (testing "we never insert any duplicates"
+                      (is (= 1 (t2/count Setting search-col search-value))))
+
+                    (testing "later calls just return the existing row as well"
+                      (is (= latest (thunk)))
+                      (is (= 1 (t2/count Setting search-col search-value)))))
+
+                :value
+                (do
+                  (testing "there may be race conditions, but we insert at least once"
+                    (is (pos? n)))
+
+                  (testing "we returned the same values that were inserted into the database"
+                    (is (= results (set (t2/select Setting search-col search-value)))))
+
+                  (testing "later calls just return an existing row as well"
+                    (is (contains? results (thunk)))
+                    (is (= results (set (t2/select Setting search-col search-value))))))))
+
+            ;; Since we couldn't use with-temp, we need to clean up manually.
+            (finally
+              (t2/delete! Setting search-col search-value))))))))
```

Comment:
Did sanya get around to writing that `with-model-cleanup` helper? Could come in use here

Reply:
@piranha are you my knight with shining macro?

Reply:
I could be, I just need to sort out that hooks madness!

Reply:
I don't want to couple this test to that, as the fact that we're using that model is incidental, and we do not have a corresponding `defsetting` registration.